### PR TITLE
Use primitive classes or instances in primitive options

### LINKED ISSUE #980: I created Multiple Input Primitives,but the primitive_options is invalid.
### I created Multiple Input Primitives,but the primitive_options is invalid.

-----
*Issues created here on Github are for bugs or feature requests. For usage questions and questions about errors, please ask on Stack Overflow with the [featuretools](https://stackoverflow.com/questions/tagged/featuretools) tag. Check the [documentation](https://docs.featuretools.com/en/stable/help.html) for further guidance on where to ask your question.*

#### Bug/Feature Request Description

I created Multiple Input Primitives,but the primitive_options is invalid.
It has 3 inputs and 5 ouputs.
When primitive_options={miss_reject_phone_count: {'ignore_variables': {'calllog': ['number']}}},
the output is not ingore calllog.number.
And When primitive_options={
                                           miss_reject_phone_count: [{'include_variables': {'calllog': ['name']}},
                                                     {'include_variables': {'calllog': ['type']}},
                                                     {'ignore_variables': {'calllog': ['duration']}}]},
have a error 
image

image

#### Output of ``featuretools.show_info()``

<details>
image

[paste the output of ``featuretools.show_info()`` here below this line]
2020-05-19 19:17:00,004 featuretools - WARNING    Featuretools failed to load plugin nlp_primitives from library nlp_primitives.__init__. For a full stack trace, set logging to debug.
Featuretools version: 0.13.4
Featuretools installation directory: /usr/local/lib/python3.6/site-packages/featuretools

SYSTEM INFO
-----------
python: 3.6.8.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-1062.18.1.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: zh_CN.UTF-8
LOCALE: zh_CN.UTF-8

INSTALLED VERSIONS
------------------
numpy: 1.18.4
pandas: 0.25.3
tqdm: 4.44.1
PyYAML: 5.3.1
cloudpickle: 1.3.0
dask: 2.14.0
distributed: 2.14.0
psutil: 5.7.0
Click: 7.0
pip: 20.1
setuptools: 46.2.0
</details>

### PR DESCRIPTION
**Pull request by frances-h**
- allow use of primitive classes or primitive instances as keys in `primitive_options`
- primitive instance will only use its options, even if generic options for the primitive class are present in the options dictionary
- update the FAQ and tutorial to reflect changes

Resolves #980

### CODE REVIEW

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py [hunk 1]
```diff
@@ -1195,3 +1208,84 @@ def test_primitive_options_multiple_inputs(es):
...
+                                    trans_primitives=[],
+                                    primitive_options=options1)
+    features3 = dfs_obj3.build_features()
+
+    dfs_obj4 = DeepFeatureSynthesis(target_entity_id='cohorts',
+                                    entityset=es,
+                                    agg_primitives=[Mean],
+                                    trans_primitives=[],
+                                    primitive_options=options2)
+    features4 = dfs_obj4.build_features()
```
Comment (thehomebrewnerd) [inline featuretools/tests/synthesis/test_deep_feature_synthesis.py:1261]: Maybe we simplify this a little by looping over the four options. Something like this:
```
primitives = [Mean, 'mean']
options = [options1, options2]
results = []
for primitive in primitives:
    for option in options:
        features.append(DeepFeatureSynthesis(...).build_features())

```
And then update the assertions that follow accordingly to use the items in the results list as needed.

File: docs/source/guides/specifying_primitive_options.rst [hunk 2]
```diff
@@ -54,6 +54,9 @@ Specifying for Individual Primitives
 Options for individual primitives or groups of primitives are set by the ``primitive_options`` parameter of DFS. This parameter
 maps any desired options to specific primitives. In the case of conflicting options, options set at this level will override
 options set at the entire DFS run level, and the include options will always take priority over their ignore counterparts.
+Using the string primitive name or the primitive type will apply the options to all primitives of the same name. You can
+also set options for a specific instance of a primitive. Note, however, that specifying options for a specific instance will
```
Comment (thehomebrewnerd) [inline docs/source/guides/specifying_primitive_options.rst:58]: If I am understanding correctly, the only way to set the options for a specific instance that are different than the options for the primitive generally is to use the instance as the key. If that is correct, we might want to update this a bit for clarity. Maybe change `...also set options for a specific instance of a primitive.` to `...also set options for a specific instance of a primitive by using the primitive instance as a key in the ``primitive_options`` dictionary.`

File: featuretools/tests/synthesis/test_deep_feature_synthesis.py [hunk 3]
```diff
@@ -1195,3 +1208,84 @@ def test_primitive_options_multiple_inputs(es):
...
+                                   primitive_options=options)
+
+    features = dfs_obj.build_features()
+    for f in features:
+        deps = f.get_dependencies(deep=True)
+        entities = [d.entity.id for d in deps]
+        if f.primitive == skipna_mean:
+            assert all(entity == 'stores' for entity in entities)
+        elif isinstance(f.primitive, Mean):
+            assert 'stores' not in entities
```
Comment (thehomebrewnerd) [inline featuretools/tests/synthesis/test_deep_feature_synthesis.py:1291]: Should we also add a test to make sure an error is raised if conflicting options are specified using a combination of primitive type and string name? Seems to be working correctly, but I don't think we explicitly verify that in our tests.

File: featuretools/primitives/options_utils.py [hunk 4]
```diff
@@ -53,14 +56,17 @@ def generate_all_primitive_options(all_primitives,
...
     # for now, only use primitive names as option keys
     for primitive in all_primitives:
-        if not isinstance(primitive, str):
-            primitive = primitive.name
-        if primitive in primitive_options:
+        if primitive in primitive_options and primitive.name in primitive_options:
+            msg = "Options present for primitive instance and generic " \
+                  "primitive class (%s), primitive instance will not use generic " \
+                  "options" % (primitive.name)
+            logger.warning(msg)
```
Comment (thehomebrewnerd) [inline featuretools/primitives/options_utils.py:63]: I don't think we verify the warning message is logged as expected. Maybe we can update `test_primitive_options_instantiated_primitive` to check the warning is raised properly as well.

File: docs/source/changelog.rst [hunk 5]
```diff
@@ -21,7 +21,8 @@ Changelog
         * Allow cutoff time dataframe columns to be in any order (:pr:`969`, :pr:`995`)
         * Add Age primitive, and make it a default transform primitive for DFS (:pr:`987`)
         * Add ``include_cutoff_time`` arg - control whether data at cutoff times are included in feature calculations (:pr:`959`)
-        * Allow ``variables_types`` to be referenced by their ``type_string``
+        * Add ability to use primitive classes and instances as keys in primitive_options dictionary (:pr:`993`)
```
Comment (rwedge) [inline docs/source/changelog.rst:24]: Needs to be moved to the Future Release section

**Pull request merged**